collector/zfs.go

```diff
@@ -0,0 +1,157 @@
+package collector
+
+// +build !nofilesystem
+// +build !nozfs
+
+import (
+	"errors"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+)
+
+type zfsMetricValue int
+
+const zfsErrorValue = zfsMetricValue(-1)
+
+var zfsNotAvailableError = errors.New("ZFS / ZFS statistics are not available")
+
+type zfsSysctl string
+type zfsSubsystemName string
+
+const (
+	arc = zfsSubsystemName("zfs_arc")
+)
+
+//------------------------------------------------------------------------------
+//                                    Metrics
+//------------------------------------------------------------------------------
+
+type zfsMetric struct {
+	subsystem zfsSubsystemName // The Prometheus subsystem name
+	name      string           // The Prometheus name of the metric
+	sysctl    zfsSysctl        // The sysctl of the ZFS metric
+}
+
+func NewZFSMetric(subsystem zfsSubsystemName, sysctl, name string) zfsMetric {
+	return zfsMetric{
+		sysctl:    zfsSysctl(sysctl),
+		subsystem: subsystem,
+		name:      name,
+	}
+}
+func (m *zfsMetric) BuildFQName() string {
+	return prometheus.BuildFQName(Namespace, string(m.subsystem), m.name)
+}
+
+func (m *zfsMetric) HelpString() string {
+	return m.name
+}
+
+//------------------------------------------------------------------------------
+//                                 Collector
+//------------------------------------------------------------------------------
+
+func init() {
+	Factories["zfs"] = NewZFSCollector
+}
+
+type zfsCollector struct {
+	zfsMetrics     []zfsMetric
+	metricProvider zfsMetricProvider
+}
+
+func NewZFSCollector() (Collector, error) {
+	err := zfsInitialize()
```

Comment:
Right now it is.
Let me explain why it's still there: the FreeBSD implementation could be made even faster if the sysctls strings were mapped to MIBs (integer array defining the sysctl). This is has to be done only once on initialization.

However, with recent changes (no errors on missing ZFS module) this is no longer possible.

Will remove.